domain/port/service/watcher.go

```diff
@@ -0,0 +1,168 @@
+		mapper eventsource.Mapper,
+	) (watcher.StringsWatcher, error)
+
+	// NewNamespaceNotifyMapperWatcher returns a new namespace notify watcher
+	// for events based on the input change mask and mapper.
+	NewNamespaceNotifyMapperWatcher(
+		namespace string, changeMask changestream.ChangeType, mapper eventsource.Mapper,
+	) (watcher.NotifyWatcher, error)
+}
+
+// WatcherState describes the methods that the service needs for its watchers.
+type WatcherState interface {
+	// WatchOpenedPortsTable returns the name of the table that should be watched
+	WatchOpenedPortsTable() string
+
+	// InitialWatchOpenedPortsStatement returns the query to load the initial event
+	// for the WatchOpenedPorts watcher
+	InitialWatchOpenedPortsStatement() string
+
+	// GetMachinesForEndpoints returns map from endpoint uuids to the uuids of
+	// the machines which host that endpoint for each provided endpoint uuid.
+	GetMachinesForEndpoints(ctx context.Context, endpointUUIDs []string) ([]string, error)
+
+	FilterEndpointsForApplication(ctx context.Context, app string, eps []string) (set.Strings, error)
+}
+
+// WatchOpenedPorts returns a strings watcher for opened ports. This watcher
+// emits events for changes to the opened ports table. Each emitted event
+// contains the machine UUID which is associated with the port range.
+func (s *WatchableService) WatchOpenedPorts(ctx context.Context) (watcher.StringsWatcher, error) {
+	return s.watcherFactory.NewNamespaceMapperWatcher(
+		s.st.WatchOpenedPortsTable(),
+		changestream.Create|changestream.Delete,
+		eventsource.InitialNamespaceChanges(s.st.InitialWatchOpenedPortsStatement()),
+		s.endpointToMachineMapper,
+	)
+}
+
+// WatchOpenedPortsForApplication returns a notify watcher for opened ports. This
+// watcher emits events for changes to the opened ports table that are associated
+// with the given application
+func (s *WatchableService) WatchOpenedPortsForApplication(ctx context.Context, applicationUUID string) (watcher.NotifyWatcher, error) {
+	return s.watcherFactory.NewNamespaceNotifyMapperWatcher(
+		s.st.WatchOpenedPortsTable(),
+		changestream.Create|changestream.Delete,
+		s.filterForApplication(applicationUUID),
+	)
+}
+
+// endpointToMachineMapper is an eventsource.Mapper that maps a slice of
+// changestream.ChangeEvent containing endpoint UUIDs to a slice of
+// changestream.ChanegEvent containing the UUIDs of the machines the corresponding
+// endpoint is located on.
+func (s *WatchableService) endpointToMachineMapper(
+	ctx context.Context, db database.TxnRunner, events []changestream.ChangeEvent,
+) ([]changestream.ChangeEvent, error) {
+
+	endpointUUIDs := transform.Slice(events, func(e changestream.ChangeEvent) string {
+		return e.Changed()
+	})
+
+	machineUUIDs, err := s.st.GetMachinesForEndpoints(ctx, endpointUUIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	newEvents := make([]changestream.ChangeEvent, len(machineUUIDs))
+	for i, machineUUID := range machineUUIDs {
```

Comment:
**nitpick:** I tend to prefers the above way:

```go

	newEvents := make([]changestream.ChangeEvent, 0, len(machineUUIDs)) // init with 0 size + capacity
	for _, machineUUID := range machineUUIDs {
Comment
		newEvents = append(newEvents, changeEventShim{
			changeType: changestream.Update,
			namespace:  "machine",
			changed:    machineUUID,
		})
	}
```

* Performance-wise it is the same (as far i tested a couple of months ago, it shouldn't have change), since there is no realloc for the slice.
* It is more future proof: 
  * if for some reason `GetMachinesForEndpoints` return a map instead of a slice, no need to change the code
  * if for some reason we decide to filter out some uuid inside the loop, no risk of empty ChangeEvent being generated in the loop.
* Readability-wise, I find it a bit better : you avoid a variable (i), which is painful to remember (ok, not in this case, but... Future!)

Reply:
yeah fair. I don't really find one more or less readable myself, but happy to change over